Python/ceval.c

```diff
@@ -4211,7 +4211,35 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
             CHECK_EVAL_BREAKER();
             DISPATCH();
         }
+        case TARGET(CALL_METHOD_KW): {
+            /* Designed to work in tandem with LOAD_METHOD. Same as CALL_METHOD
+            but pops TOS to get a tuple of keyword names. */
+            PyObject **sp, *res, *meth;
+            PyObject *names = NULL;
+
+            names = POP();
+
+            sp = stack_pointer;
+
+            meth = PEEK(oparg + 2);
+            if (meth == NULL) {
```

Comment:
Given that `call_function` is inlined and this branch might not be that predictable, it might be an idea to use a branch-less form here:
```C
int meth = PEEK(oparg + 2) != NULL;
res = call_function(tstate, &trace_info, &sp, oparg + meth, names);
stack_pointer = sp;
STACK_SHRINK(1-meth);
```


```

Reply:
Thanks for the suggestion - that's really clever! This branch has been bugging me ever since I took the code from ``CALL_METHOD``. But ``STACK_SHRINK`` slipped my mind so it eluded me for a while.

BTW, should I apply this to ``CALL_METHOD`` as well since their code is basically identical (apart from a ``POP``)? Or leave that for another PR?

Reply:
Crazy story:
I was getting an assertion error at interpreter start: ``Assertion failed: EMPTY(), file ~\GitHub\cpython\Python\ceval.c, line 2398``
The offending line is in here:
```C
#define EMPTY()           (STACK_LEVEL() == 0)
...
        case TARGET(RETURN_VALUE): {
            retval = POP();
            assert(EMPTY());
            f->f_state = FRAME_RETURNED;
            f->f_stackdepth = 0;
            goto exiting;
        }
```
I narrowed it down to the ``STACK_SHRINK(1-meth_found)`` line; replacing that with a branch fixes it. But I didn't understand why that's causing issues: ``POP()`` and ``STACK_SHRINK`` macros expand to quite similar things.

Then I suddenly found out: ``STACK_SHRINK(n)`` expands to ``BASIC_STACKADJ(-n)``. Note the lack of a parentheses around ``-n``. So ``STACK_SHRINK(1 - meth_found)`` expanded to ``BASIC_STACKADJ(-1 - meth_found)`` :-(.

I'm amazed that out of all the macros to adjust the stack, this was the only one with that problem (note the rest has the parentheses):
```C
#define STACK_LEVEL()     ((int)(stack_pointer - f->f_valuestack))
#define EMPTY()           (STACK_LEVEL() == 0)
#define TOP()             (stack_pointer[-1])
#define SECOND()          (stack_pointer[-2])
#define THIRD()           (stack_pointer[-3])
#define FOURTH()          (stack_pointer[-4])
#define PEEK(n)           (stack_pointer[-(n)])
#define SET_TOP(v)        (stack_pointer[-1] = (v))
#define SET_SECOND(v)     (stack_pointer[-2] = (v))
#define SET_THIRD(v)      (stack_pointer[-3] = (v))
#define SET_FOURTH(v)     (stack_pointer[-4] = (v))
#define BASIC_STACKADJ(n) (stack_pointer += n)
#define BASIC_PUSH(v)     (*stack_pointer++ = (v))
#define BASIC_POP()       (*--stack_pointer)
```
Guess my next PR is to add the parentheses so others don't trip over the same thing.